windows/CodePush.cs

```diff
@@ -0,0 +1,668 @@
+            return pendingUpdate != null &&
+                    !(bool)pendingUpdate[PENDING_UPDATE_IS_LOADING_KEY] &&
+                    (packageHash == null || ((string)pendingUpdate[PENDING_UPDATE_HASH_KEY]).Equals(packageHash));
+        }
+
+        private void RemoveFailedUpdates()
+        {
+            ApplicationDataContainer settings = GetCodePushSettings();
+            settings.Values.Remove(FAILED_UPDATES_KEY);
+        }
+
+        private void RemovePendingUpdate()
+        {
+            ApplicationDataContainer settings = GetCodePushSettings();
+            settings.Values.Remove(PENDING_UPDATE_KEY);
+        }
+
+        private async Task RollbackPackage()
+        {
+            JObject failedPackage = await codePushPackage.GetCurrentPackage();
+            SaveFailedUpdate(failedPackage);
+            await codePushPackage.RollbackPackage();
+            RemovePendingUpdate();
+        }
+
+        private void SaveFailedUpdate(JObject failedPackage)
+        {
+            ApplicationDataContainer settings = GetCodePushSettings();
+            string failedUpdatesString = (string)settings.Values[FAILED_UPDATES_KEY];
+            JArray failedUpdates;
+            if (failedUpdatesString == null)
+            {
+                failedUpdates = new JArray();
+            }
+            else
+            {
+                failedUpdates = JArray.Parse(failedUpdatesString);
+            }
+
+            failedUpdates.Add(failedPackage);
+            settings.Values[FAILED_UPDATES_KEY] = JsonConvert.SerializeObject(failedUpdates);
+        }
+
+        private void SavePendingUpdate(string packageHash, bool isLoading)
+        {
+            ApplicationDataContainer settings = GetCodePushSettings();
+            JObject pendingUpdate = new JObject();
+            pendingUpdate[PENDING_UPDATE_HASH_KEY] = packageHash;
+            pendingUpdate[PENDING_UPDATE_IS_LOADING_KEY] = isLoading;
+            settings.Values[PENDING_UPDATE_KEY] = JsonConvert.SerializeObject(pendingUpdate);
+        }
+
+        public static bool IsUsingTestConfiguration()
+        {
+            return testConfigurationFlag;
+        }
+
+        public static void SetUsingTestConfiguration(bool shouldUseTestConfiguration)
+        {
+            testConfigurationFlag = shouldUseTestConfiguration;
+        }
+
+        public async Task ClearUpdates()
+        {
+            await codePushPackage.ClearUpdates();
+            RemovePendingUpdate();
+            RemoveFailedUpdates();
+        }
+
+        private class CodePushNativeModule : ReactContextNativeModuleBase
+        {
+            private ILifecycleEventListener lifecycleEventListener = null;
+            private int minimumBackgroundDuration = 0;
+            private ReactContext reactContext;
+
+            private class CodePushResumeListener : ILifecycleEventListener
+            {
+                private DateTime? lastPausedDate = null;
```

Comment:
may need to do that on the cordova side as well, I think I named the variable there "lastPausedDate" to follow this lol